[FEATURE] 채팅 서비스 base of base 코드 작성

hyun98 opened the pull request:
close #2

# 변경 사항
- chat 패키지 생성

# 참고 사항
- 날짜 라이브러리는 jodatime 사용

hyun98 committed:
create chat service base code #2

jojaeng2 commented on sns_service/build.gradle.kts:
```diff
@@ -29,6 +29,11 @@ dependencies {
+	implementation ("org.springframework.boot:spring-boot-starter-data-redis")
+
+	// https://mvnrepository.com/artifact/joda-time/joda-time
+	implementation("joda-time:joda-time:2.12.5")
```
-> datetime 동시성 문제 해결

gusah009 replied:
`joda-time`이 어떤 동시성 문제를 해결했는지 알 수 있을까요?

[Java8의 `LocalDateTime` vs `JodaTime` 스택 오버플로우 글](https://stackoverflow.com/questions/24631909/differences-between-java-8-date-time-api-java-time-and-joda-time)을 봤을 땐 그냥 `LocalDateTime` 사용이 더 좋은 것 같아서용

[더 자세한 비교](http://time4j.net/tutorial/appendix.html)

jojaeng2 replied:
@hyun98 얘가 그랬어요.

hyun98 replied:
작년에 K 인턴할때 들은 내용이었는데, 제가 착각을 했네여
그때 자바의 `Date()` 가 스레드 세이프 하지 않다고 들었고 jodatime 을 사용하면 된다고 들었거든요 `Date()` 랑 LocalDateTime 은 다른건데,,ㅋㅋㅋ

스택 오버플로우 글이랑 다른 자료들 보니까 자바8 이후로 java.time 패키지는 jodatime의 개념을 차용해서 동시성 문제를 해결하고 시간 데이터를 안정적으로 제공하는 것 같아용
결론은 표준이 되어버린 java.time 패키지 씁시다

jojaeng2 commented on sns_service/src/main/kotlin/joryu/sns_service/chat/config/ChatRedisConfig.kt:
```diff
@@ -0,0 +1,17 @@
+
+    @Bean
+    fun redisTemplate(redisConnectionFactory: RedisConnectionFactory): RedisTemplate<String, String> {
+        val template = RedisTemplate<String, String>()
```
이거 전역으로 빼면 안되나요??

hyun98 replied:
이후 Redis 연동 작업 시 반영하겠습니다

jojaeng2 commented on sns_service/src/main/kotlin/joryu/sns_service/chat/controller/ChatController.kt:
```diff
@@ -0,0 +1,32 @@
+) {
+    @PostMapping("/api/chat")
+    fun createMessage(@RequestBody newMessage: NewMessageRequest): ResponseEntity<Any> {
+        if (newMessage.message.isEmpty()) {
+            return ResponseEntity.badRequest().body("Empty Message now Allow")
+        }
```
message가 isEmpty인 경우를 controller단에서 확인하지 말고, 아예 RequestBody가 안맞으니깐 controller까지 오기 전에 걸러지는게 맞는 것 같음.

message에 대해서 nullable을 허용하지 않으면 해결될 문제같음.

hyun98 replied:
input request 에서 message 필드에 대해서 nullable 설정했습니다람쥐

jojaeng2 commented on sns_service/src/main/kotlin/joryu/sns_service/chat/controller/ChatController.kt:
```diff
@@ -0,0 +1,32 @@
+        return ResponseEntity.ok().build()
+    }
+
+    @GetMapping("api/chat")
```
api앞에 '/' 놓침

jojaeng2 commented on sns_service/src/main/kotlin/joryu/sns_service/chat/service/ChatService.kt:
```diff
@@ -0,0 +1,25 @@
+    private val chatRepository: ChatRepository,
+    private val profileRepository: ProfileRepository
+){
+
```
@Transaction(readOnly=true) 추가

jojaeng2 commented on sns_service/src/main/kotlin/joryu/sns_service/chat/service/ChatService.kt:
```diff
@@ -0,0 +1,25 @@
+    fun findAll(): List<ChatMessage?> {
+        return chatRepository.findAll()
+    }
+
```
@Transaction 추가

jojaeng2 commented on sns_service/src/main/kotlin/joryu/sns_service/chat/service/ChatService.kt:
```diff
@@ -0,0 +1,25 @@
+@Service
+@RequiredArgsConstructor
+class ChatService(
+    private val chatRepository: ChatRepository,
+    private val profileRepository: ProfileRepository
```
private good good

jojaeng2 reviewed (commented):
**LGTM**

hyun98 committed:
fix error code

1. add service transactional annotation
2. add typo ChatMessage 'val'
3. fix nullable field request message
